### PR TITLE
Test optional scope parameter in token refresh sequence

### PR DESCRIPTION
**Pull request by arscan**
Addresses #246 which was a request for improved testing on our token refresh.  We also made a number of changes, including:
* Condensed what used to be three tests into a single test, because we needed to duplicate that single test to handle multiple cases.  It was far too repetitive to keep the same test structure.
* Identified a number of minor issues in the SMART spec that went unnoticed earlier.  We had to change some of our tests to accommodate that (noted in the referenced issue).
* Significantly bulked up our self testing

You can try out these changes by executing out the refresh token sequence on any server

### CODE REVIEW

File: test/sequence/token_refresh_sequence_test.rb [hunk 1]
```diff
@@ -7,82 +7,162 @@
...
+
+    headers = {
+      'Content-Type' => 'application/x-www-form-urlencoded'
+    }
+    body = {
+      'grant_type' => 'refresh_token',
+      'refresh_token' => @instance.refresh_token
+    }
+
+    body_with_scope = {
```
Comment (Jammjammjamm) [inline test/sequence/token_refresh_sequence_test.rb]:
```
body_with_scope = body.merge('scope' => @instance.scopes)
```

File: test/sequence/token_refresh_sequence_test.rb [hunk 2]
```diff
@@ -7,82 +7,162 @@
...
-    # To test rejection of invalid client_id
+    response_headers = { content_type: 'application/json; charset=UTF-8',
+                         cache_control: 'no-store',
+                         pragma: 'no-cache' }
+
+    response_headers.delete(:cache_control) if failure_mode == :cache_control_off
+    response_headers.delete(:pragma) if failure_mode == :pragma_off
+
+    exchange_response_json = exchange_response.to_json
+    exchange_response_json = '<bad>' if failure_mode == :bad_json_response
```
Comment (Jammjammjamm) [inline test/sequence/token_refresh_sequence_test.rb]: I think it would be nice to group all of the `failure_mode` changes together.

File: test/sequence/token_refresh_sequence_test.rb [hunk 3]
```diff
@@ -7,82 +7,162 @@
...
+  end
+
+  def test_fail_if_bad_json_response
+    setup_mocks(:bad_json_response)
+
+    sequence_result = @sequence.start
+    assert sequence_result.fail?
+  end
+
+  def test_fail_if_scope_must_be_in_payload
```
Comment (Jammjammjamm) [inline test/sequence/token_refresh_sequence_test.rb]: Do we need a corresponding test for if the scope must be excluded from the payload?

File: lib/app/modules/smart/token_refresh_sequence.rb [hunk 4]
```diff
@@ -28,133 +31,150 @@ class TokenRefreshSequence < SequenceBase
...
-        @token_response_body = JSON.parse(@token_response.body)
+        token_response = perform_refresh_request(INVALID_CLIENT_ID, @instance.refresh_token)
+        assert_response_bad_or_unauthorized token_response
+      end
 
-        assert @token_response_body.key?('access_token'), 'Token response did not contain access_token as required'
+      def validate_and_save_refresh_response(token_response)
+        assert_response_ok(token_response)
+        assert_valid_json(token_response.body)
+        token_response_body = JSON.parse(token_response.body)
```
[the remark this answers is not in the document]
Comment (arscan) [reply lib/app/modules/smart/token_refresh_sequence.rb]: Should it?  Yeah, it is a bit wordy to assert then parse instead of doing it in one step, but having an `assert_*` provide more functionality than making an assertion seems strange to me.

File: lib/app/modules/smart/token_refresh_sequence.rb [hunk 5]
```diff
@@ -28,133 +31,150 @@ class TokenRefreshSequence < SequenceBase
...
+
+        assert token_response_body.key?('access_token'), 'Token response did not contain access_token as required'
```
[the remark this answers is not in the document]
Comment (arscan) [reply lib/app/modules/smart/token_refresh_sequence.rb]: Good catch on `expires_in`.

Regarding moving the code, I would agree on moving that up here from a code clarity perspective.  But we do it this way because we often run into situations where systems are mostly correct, but slightly wrong.  And it is frustrating for developers to be blocked from see how conformant they are on steps 2, 3, and 4 if they need to be 100% conformant on step 1 before continuing.  So we check and save the most important part (the bearer token), then look at all the other pieces of information that are required but that we don't actually use.   I'll see if I can make that more clear.

File: lib/app/modules/smart/token_refresh_sequence.rb [hunk 6]
```diff
@@ -28,133 +31,150 @@ class TokenRefreshSequence < SequenceBase
...
+        scopes = token_response_body['scope'] || @instance.scopes
 
         @instance.save!
         @instance.update(scopes: scopes)
 
-        if @token_response_body.key?('id_token')
+        if token_response_body.key?('id_token')
           @instance.save!
-          @instance.update(id_token: @token_response_body['id_token'])
+          @instance.update(id_token: token_response_body['id_token'])
```
[the remark this answers is not in the document]
Comment (arscan) [reply lib/app/modules/smart/token_refresh_sequence.rb]: There shouldn't be, but DataMapper is fiddly about synchronized cached data and I'm concerned that removing this will cause bad behavior.  I would like to factor out DataMapper references in our test dsl and just think about it in terms of a state/context object that the tests can interact with, which is written so we don't have to ever do the 'save!' thing.

File: lib/app/modules/smart/token_refresh_sequence.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (arscan) [reply lib/app/modules/smart/token_refresh_sequence.rb]: Interesting how `expires_in` is required here but just `recommended` when getting your access token the first time...

File: lib/app/modules/smart/token_refresh_sequence.rb [hunk 7]
```diff
@@ -28,133 +31,150 @@ class TokenRefreshSequence < SequenceBase
...
         end
 
-        if @token_response_body.key?('refresh_token')
+        if token_response_body.key?('refresh_token')
           @instance.save!
-          @instance.update(refresh_token: @token_response_body['refresh_token'])
+          @instance.update(refresh_token: token_response_body['refresh_token'])
+        end
+        [:cache_control, :pragma].each do |key|
```
[the remark this answers is not in the document]
Comment (arscan) [reply lib/app/modules/smart/token_refresh_sequence.rb]: Great question, I'll see if i can track that down.

File: lib/app/modules/smart/token_refresh_sequence.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (Jammjammjamm) [reply lib/app/modules/smart/token_refresh_sequence.rb]: This is the assertion code:
```
    def assert_valid_json(json)
      JSON.parse(json)
    rescue JSON::ParserError
      raise AssertionException, 'Invalid JSON'
    end
```
I think validating json is always annoying because you validate it by parsing it, so any code like `data = parse(json) if valid(json)` is really just parsing the json twice. But if you'd prefer to treat the assertion as just an assertion rather a "parse the json but raise a special error if it's invalid" method, and leave the code as is that's fine with me.

[the remark this answers is not in the document]
Comment (arscan) [reply lib/app/modules/smart/token_refresh_sequence.rb]: Yeah, understood, i could go either way.  I'm just going to leave this as-is.  Thanks for pointing this out though.

File: lib/app/modules/smart/token_refresh_sequence.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (arscan) [reply lib/app/modules/smart/token_refresh_sequence.rb]: Moved to required as confirmed in https://github.com/siteadmin/inferno/issues/246#issuecomment-514692915

I opened a ticket for us to provide a warning when expires_in is not provided during the launch sequences.

File: lib/app/modules/smart/token_refresh_sequence.rb [hunk 8]
```diff
@@ -17,9 +17,12 @@ class TokenRefreshSequence < SequenceBase
       Refresh tokens are typically longer lived than access tokens and allow client applications to obtain a new access token
       Refresh tokens themselves cannot provide access to resources on the server.
 
+      Token refreshes are accomplished through a `POST` request to the token exchange endpoint as described in the
+      [SMART App Launch Framework](http://www.hl7.org/fhir/smart-app-launch/#step-5-later-app-uses-a-refresh-token-to-obtain-a-new-access-token)
+
       # Test Methodology
 
-      Inferno will attempt to exchange the refresh token for a new access token and verify that the information returned
+      This test attempt to exchange the refresh token for a new access token and verify that the information returned
```
[the remark this answers is not in the document]
Comment (arscan) [reply lib/app/modules/smart/token_refresh_sequence.rb]: Fixed, good catch.

File: lib/app/modules/smart/token_refresh_sequence.rb (hunk 7 above)
[the remark this answers is not in the document]
Comment (arscan) [reply lib/app/modules/smart/token_refresh_sequence.rb]: Excellent catch, moved to a warning as recommended here https://github.com/siteadmin/inferno/issues/246#issuecomment-514692915

File: test/sequence/token_refresh_sequence_test.rb (hunk 1 above)
Comment (arscan) [reply test/sequence/token_refresh_sequence_test.rb]: I figured there would be a better way to do this.  Changed.

File: test/sequence/token_refresh_sequence_test.rb (hunk 3 above)
Comment (arscan) [reply test/sequence/token_refresh_sequence_test.rb]: Added another test.

File: test/sequence/token_refresh_sequence_test.rb (hunk 2 above)
Comment (arscan) [reply test/sequence/token_refresh_sequence_test.rb]: Consolidated all of these into one spot.

**Comment by arscan** [comment]
I believe I've addressed all the comments.  All excellent points.
